domain/storage/service/service.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/juju/collections/transform"
+	"github.com/juju/errors"
+
+	domainstorage "github.com/juju/juju/domain/storage"
+	storageerrors "github.com/juju/juju/domain/storage/errors"
+	"github.com/juju/juju/internal/storage"
+)
+
+// State defines an interface for interacting with the underlying state.
+type State interface {
+	CreateStoragePool(ctx context.Context, pool domainstorage.StoragePoolDetails) error
+	DeleteStoragePool(ctx context.Context, name string) error
+	ReplaceStoragePool(ctx context.Context, pool domainstorage.StoragePoolDetails) error
+	ListStoragePools(ctx context.Context, filter domainstorage.StoragePoolFilter) ([]domainstorage.StoragePoolDetails, error)
+	GetStoragePoolByName(ctx context.Context, name string) (domainstorage.StoragePoolDetails, error)
+}
+
+// Logger facilitates emitting log messages.
+type Logger interface {
+	Debugf(string, ...interface{})
+}
+
+// Service defines a service for interacting with the underlying state.
+type Service struct {
+	st       State
+	logger   Logger
+	registry storage.ProviderRegistry
+}
+
+// NewService returns a new Service for interacting with the underlying state.
+func NewService(st State, logger Logger, registry storage.ProviderRegistry) *Service {
+	return &Service{
+		st:       st,
+		logger:   logger,
+		registry: registry,
+	}
+}
+
+// PoolAttrs define the attributes of a storage pool.
+type PoolAttrs map[string]any
```

Comment:
You know this code better than me. But should we rip the band aid off now and just make this [string]string and get some type safety back in our lives?

It is up to the provider validation logic to infer meaning.

Reply:
the args come over the wire as any (as supplied by the cli and client callers). the facade just hands the args as received to the service. the service is responsible for validation and coercion.

Reply:
the facade could coerce to strings off the wire.

Reply:
no worries. Thank you for answering. I love it when the client send `any` over the wire. What could go wrong.